File: backend-api/app/schemas/project.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional, Literal


# ── Tipos anidados ────────────────────────────────────────

class TechCategory(BaseModel):
    """Una categoría del tech stack con sus tecnologías."""
    category: str
    technologies: list[str]


class Feature(BaseModel):
    """Una feature clave del proyecto."""
    title: str
    description: str


class AdjacentProject(BaseModel):
    """Referencia mínima para navegación prev/next."""
    title: str
    slug: str


class AdjacentResponse(BaseModel):
    prev: Optional[AdjacentProject] = None
    next: Optional[AdjacentProject] = None


# ── Esquema base ──────────────────────────────────────────

class ProjectBase(BaseModel):

    # ── Identificación
    slug: str
    order: int = 0
    project_type: Literal["web_app", "mobile_app", "api", "library", "tool"] = "web_app"
    is_published: bool = False
    is_featured: bool = False

    # ── Hero
    title: str
    tagline: Optional[str] = None
    hero_image_url: Optional[str] = None
    cta_primary_label: Optional[str] = None
    cta_primary_url: Optional[str] = None
    cta_secondary_label: Optional[str] = None
    cta_secondary_url: Optional[str] = None
    repository_url: Optional[str] = None

    # ── Metadata Strip
    role: Optional[str] = None
    year: Optional[int] = None
    status: Optional[Literal["Live", "In Development", "Deprecated", "Open Source"]] = None
    client: Optional[str] = None
    duration: Optional[str] = None
    team_size: Optional[str] = None

    # ── About / Challenge
    challenge_paragraph_1: Optional[str] = None
    challenge_paragraph_2: Optional[str] = None
    key_challenges: list[str] = []

    # ── Tech Stack y Features
    tech_stack: list[TechCategory] = []
    features: list[Feature] = []

    # ── Métricas
    metrics: list[str] = []

    # ── Galería
    main_screenshot_url: Optional[str] = None
    screenshots: list[str] = []

    # ── Card del listado
    card_thumbnail_url: Optional[str] = None
    card_short_description: Optional[str] = None
    tags: list[str] = []

    # ── SEO
    seo_title: Optional[str] = None
    seo_description: Optional[str] = None


# ── Modelos de entrada/salida ─────────────────────────────

class ProjectCreate(ProjectBase):
    pass


class ProjectResponse(ProjectBase):
    id: int
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

File: desktop-app/src/types/project.ts
// ── Tipos anidados ─────────────────────────────────────────────────────────────

export interface TechCategory {
  category: string;
  technologies: string[];
}

export interface Feature {
  title: string;
  description: string;
}

// ── Proyecto completo (respuesta de la API) ────────────────────────────────────

export interface Project {
  id: number;

  // Identificación
  slug: string;
  order: number;
  project_type: "web_app" | "mobile_app" | "api" | "library" | "tool";
  is_published: boolean;
  is_featured: boolean;
  created_at: string;
  updated_at: string;

  // Hero
  title: string;
  tagline: string | null;
  hero_image_url: string | null;
  cta_primary_label: string | null;
  cta_primary_url: string | null;
  cta_secondary_label: string | null;
  cta_secondary_url: string | null;
  repository_url: string | null;

  // Metadata Strip
  role: string | null;
  year: number | null;
  status: "Live" | "In Development" | "Deprecated" | "Open Source" | null;
  client: string | null;
  duration: string | null;
  team_size: string | null;

  // About / Challenge
  challenge_paragraph_1: string | null;
  challenge_paragraph_2: string | null;
  key_challenges: string[];

  // Tech Stack y Features
  tech_stack: TechCategory[];
  features: Feature[];

  // Métricas
  metrics: string[];

  // Galería
  main_screenshot_url: string | null;
  screenshots: string[];

  // Card del listado
  card_thumbnail_url: string | null;
  card_short_description: string | null;
  tags: string[];

  // SEO
  seo_title: string | null;
  seo_description: string | null;
}

// ── Payload de creación / actualización (todos opcionales excepto title y slug) ─

export interface ProjectCreate {
  title: string;
  slug: string;
  order?: number;
  project_type?: "web_app" | "mobile_app" | "api" | "library" | "tool";
  is_published?: boolean;
  is_featured?: boolean;
  tagline?: string;
  hero_image_url?: string;
  cta_primary_label?: string;
  cta_primary_url?: string;
  cta_secondary_label?: string;
  cta_secondary_url?: string;
  repository_url?: string;
  role?: string;
  year?: number | null;
  status?: "Live" | "In Development" | "Deprecated" | "Open Source" | null;
  client?: string;
  duration?: string;
  team_size?: string;
  challenge_paragraph_1?: string;
  challenge_paragraph_2?: string;
  key_challenges?: string[];
  tech_stack?: TechCategory[];
  features?: Feature[];
  metrics?: string[];
  main_screenshot_url?: string;
  screenshots?: string[];
  card_thumbnail_url?: string;
  card_short_description?: string;
  tags?: string[];
  seo_title?: string;
  seo_description?: string;
}
